File: .reallsys/changelogGenerate.java
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

public class changelogGenerate {

    // Configure OneAPI parameters
    private static final String API_BASE_URL = System.getenv("ONEAPI_BASE_URL");
    private static final String API_KEY = System.getenv("ONEAPI_API_KEY");
    private static final String MODEL = "gpt-5-mini";

    public static void main(String[] args) {
        try {
            // Parameters: Git repository path
            String repoPath = "."; // Current directory, or specify path like "/path/to/your/repo"

            System.out.println("Fetching Git commit history...");
            String latestTag = getLatestTag(repoPath);
            System.out.println("Latest tag found: " + latestTag);  // 添加这行调试
            String commits = getGitCommitRange(repoPath, latestTag, "HEAD");

            if (commits.isEmpty()) {
                System.err.println("No commit history found");
                return;
            }

            System.out.println("\nFetched commits:\n" + commits);
            System.out.println("\nCalling OneAPI for summarization...");

            String summary = summarizeChangelog(commits);
            System.out.println("\n=== Changelog Summary ===\n" + summary);
            BufferedWriter out = new BufferedWriter(new FileWriter("CHANGELOG.md"));
            out.write(summary);
            out.close();
            System.out.println("Changelog written to CHANGELOG.md");
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    /**
     * Get Git commit history
     */
    public static String getGitCommits(String repoPath, int count, String branch)
            throws IOException, InterruptedException {
        // Use git log command with custom format
        // %h: short hash, %an: author name, %ad: author date, %s: subject
        String format = "%h | %an | %ad | %s";

        ProcessBuilder processBuilder = new ProcessBuilder(
                "git",
                "-C", repoPath,  // Specify repository path
                "log",
                "-" + count,     // Get last N commits
                branch,
                "--pretty=format:" + format,
                "--date=short"   // Date format: YYYY-MM-DD
        );

        processBuilder.redirectErrorStream(true);
        Process process = processBuilder.start();

        StringBuilder output = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }
        }

        int exitCode = process.waitFor();
        if (exitCode != 0) {
            throw new IOException("Git command failed with exit code: " + exitCode);
        }

        return output.toString().trim();
    }

    /**
     * Get commits between two references (tags, commits, branches)
     */
    public static String getGitCommitRange(String repoPath, String fromCommit, String toCommit)
            throws IOException, InterruptedException {
        String format = "%h | %an | %ad | %s";

        ProcessBuilder processBuilder = new ProcessBuilder(
                "git",
                "-C", repoPath,
                "log",
                fromCommit + ".." + toCommit,
                "--pretty=format:" + format,
                "--date=short",
                "--tags"
        );

        processBuilder.redirectErrorStream(true);
        Process process = processBuilder.start();

        StringBuilder output = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }
        }

        process.waitFor();
        return output.toString().trim();
    }

    /**
     * Call OneAPI to summarize commits (optimized based on curl command)
     */
    public static String summarizeChangelog(String commits) throws IOException {
        URL url = new URL(API_BASE_URL);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();

        try {
            conn.setRequestMethod("POST");

            // Set all request headers (based on curl command)
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setRequestProperty("Authorization", "Bearer " + API_KEY);

            conn.setDoOutput(true);
            conn.setDoInput(true);
            conn.setConnectTimeout(30000);
            conn.setReadTimeout(60000);

            String jsonPayload = buildJsonPayload(commits);

            try (OutputStream os = conn.getOutputStream()) {
                byte[] input = jsonPayload.getBytes(StandardCharsets.UTF_8);
                os.write(input, 0, input.length);
            }

            int responseCode = conn.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_OK) {
                return parseResponse(conn.getInputStream());
            } else {
                String error = readStream(conn.getErrorStream());
                throw new IOException("API request failed, status code: " + responseCode + ", error: " + error);
            }

        } finally {
            conn.disconnect();
        }
    }

    /**
     * Build JSON request payload (matching curl's --data-raw format)
     */
    private static String buildJsonPayload(String commits) {
        String escapedCommits = escapeJson(commits);
        String prompt = escapeJson(
                "Analyze the following Git commit history and generate a clean, structured changelog.\n" +
                        "Format: hash | author | date | message\n\n" +
                        "Output ONLY the categorized changelog in this exact format:\n\n" +
                        "## Features\n" +
                        "- [hash] message\n\n" +
                        "## Bug Fixes\n" +
                        "- [hash] message\n\n" +
                        "## Performance\n" +
                        "- [hash] message\n\n" +
                        "## Documentation\n" +
                        "- [hash] message\n\n" +
                        "## Other\n" +
                        "- [hash] message\n\n" +
                        "Rules:\n" +
                        "1. Do NOT include any introductory text, notes, or explanations\n" +
                        "2. Start directly with '# CHANGELOG'\n" +
                        "3. Only include categories that have commits\n" +
                        "4. Keep commit messages concise\n" +
                        "5. Use markdown format only"
        );

        return String.format("""
            {
              "model": "%s",
              "messages": [
                {
                  "role": "system",
                  "content": "You are a changelog generator. Output only the formatted changelog without any additional commentary."
                },
                {
                  "role": "user",
                  "content": "%s\\n\\nCommit History:\\n%s"
                }
              ],
              "max_completion_tokens": 100000,
              "top_p": 1,
              "presence_penalty": 0,
              "frequency_penalty": 0,
              "stream": false
            }
            """, MODEL, prompt, escapedCommits);
    }

    /**
     * Parse API response and extract content field
     */
    private static String parseResponse(InputStream is) throws IOException {
        String response = readStream(is);
        int contentIndex = response.indexOf("\"content\"");
        if (contentIndex == -1) {
            return response;
        }

        int startQuote = response.indexOf("\"", contentIndex + 10);
        int endQuote = findClosingQuote(response, startQuote + 1);

        if (startQuote != -1 && endQuote != -1) {
            String content = response.substring(startQuote + 1, endQuote);
            return unescapeJson(content);
        }

        return response;
    }

    /**
     * Read input stream content
     */
    private static String readStream(InputStream is) throws IOException {
        if (is == null) return "";

        try (BufferedReader br = new BufferedReader(
                new InputStreamReader(is, StandardCharsets.UTF_8))) {
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = br.readLine()) != null) {
                response.append(line);
            }
            return response.toString();
        }
    }

    /**
     * Escape JSON string
     */
    private static String escapeJson(String text) {
        return text.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    /**
     * Unescape JSON string
     */
    private static String unescapeJson(String text) {
        return text.replace("\\n", "\n")
                .replace("\\r", "\r")
                .replace("\\t", "\t")
                .replace("\\\"", "\"")
                .replace("\\\\", "\\");
    }

    /**
     * Find the position of unescaped closing quote
     */
    private static int findClosingQuote(String str, int start) {
        for (int i = start; i < str.length(); i++) {
            if (str.charAt(i) == '"' && str.charAt(i - 1) != '\\') {
                return i;
            }
        }
        return -1;
    }

    /**
     * Get the latest tag in the repository
     */
    public static String getLatestTag(String repoPath) throws IOException, InterruptedException {
        ProcessBuilder processBuilder = new ProcessBuilder(
                "git",
                "-C", repoPath,
                "describe",
                "--tags",
                "--abbrev=0"
        );

        processBuilder.redirectErrorStream(true);
        Process process = processBuilder.start();

        StringBuilder output = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line);
            }
        }

        int exitCode = process.waitFor();
        if (exitCode != 0) {
            throw new IOException("Git command failed with exit code: " + exitCode + ". No tags found?");
        }

        return output.toString().trim();
    }
}

File: .reallsys/scripts/changelogGenerate.py
#!/usr/bin/env python3
"""
Changelog Generator Script
Fetches Git commit history and uses AI to generate a structured changelog.
"""

import os
import sys
import subprocess
from typing import Optional
from google import genai
from google.genai import types


class ChangelogGenerator:
    """Generate changelog from Git commits using Gemini AI API"""

    def __init__(self):
        self.api_key = os.getenv("GEMINI_API_KEY")
        self.model = "gemini-3-flash-preview"

        if not self.api_key:
            raise ValueError("GEMINI_API_KEY environment variable must be set")

        # Initialize Gemini client
        self.client = genai.Client(api_key=self.api_key)

    def get_latest_tag(self, repo_path: str = ".") -> str:
        """
        Get the latest tag in the repository

        Args:
            repo_path: Path to git repository

        Returns:
            Latest tag name

        Raises:
            RuntimeError: If git command fails or no tags found
        """
        try:
            result = subprocess.run(
                ["git", "-C", repo_path, "describe", "--tags", "--abbrev=0"],
                capture_output=True,
                text=True,
                check=True
            )
            return result.stdout.strip()
        except subprocess.CalledProcessError as e:
            raise RuntimeError(f"Git command failed: {e.stderr}. No tags found?")

    def get_git_commits(self, repo_path: str = ".", count: int = 10, branch: str = "HEAD") -> str:
        """
        Get Git commit history

        Args:
            repo_path: Path to git repository
            count: Number of commits to fetch
            branch: Branch name

        Returns:
            Formatted commit history
        """
        format_str = "%h | %an | %ad | %s"

        try:
            result = subprocess.run(
                [
                    "git",
                    "-C", repo_path,
                    "log",
                    f"-{count}",
                    branch,
                    f"--pretty=format:{format_str}",
                    "--date=short"
                ],
                capture_output=True,
                text=True,
                check=True,
                encoding='utf-8'
            )
            return result.stdout.strip()
        except subprocess.CalledProcessError as e:
            raise RuntimeError(f"Git command failed: {e.stderr}")

    def get_git_commit_range(self, repo_path: str, from_commit: str, to_commit: str) -> str:
        """
        Get commits between two references (tags, commits, branches)

        Args:
            repo_path: Path to git repository
            from_commit: Starting commit/tag
            to_commit: Ending commit/tag

        Returns:
            Formatted commit history
        """
        format_str = "%h | %an | %ad | %s"

        try:
            result = subprocess.run(
                [
                    "git",
                    "-C", repo_path,
                    "log",
                    f"{from_commit}..{to_commit}",
                    f"--pretty=format:{format_str}",
                    "--date=short",
                    "--tags"
                ],
                capture_output=True,
                text=True,
                encoding='utf-8'
            )
            return result.stdout.strip()
        except subprocess.CalledProcessError as e:
            raise RuntimeError(f"Git command failed: {e.stderr}")

    def detect_breaking_changes(self, commits: str) -> bool:
        """
        Detect if there are breaking changes in commit messages

        Args:
            commits: Commit history

        Returns:
            True if breaking changes detected, False otherwise
        """
        breaking_indicators = [
            "BREAKING CHANGE:",
            "breaking change:",
            "BREAKING:",
            "breaking:",
            "!:",  # Conventional commit with breaking change marker
        ]

        # Potential breaking change patterns (case-insensitive)
        potential_breaking_patterns = [
            "migrate",
            "migration",
            "replace.*with",
            "removed.*feature",
            "deprecate",
            "rename.*package",
            "change.*api",
            "incompatible",
            "paper.*mmkv",
            "sharedpreferences.*mmkv",
        ]

        commit_lines = commits.split('\n')
        has_potential_breaking = False

        for line in commit_lines:
            line_lower = line.lower()

            # Check for explicit breaking change indicators
            for indicator in breaking_indicators:
                if indicator in line_lower:
                    return True

            # Check for conventional commit with ! (e.g., "feat!: something")
            if " | " in line:
                parts = line.split(" | ")
                if len(parts) >= 4:
                    message = parts[3]
                    if message.strip().startswith(('feat!', 'fix!', 'refactor!', 'perf!')):
                        return True

                    # Check for potential breaking patterns in commit message
                    message_lower = message.lower()
                    for pattern in potential_breaking_patterns:
                        import re
                        if re.search(pattern, message_lower):
                            has_potential_breaking = True
                            break

        return has_potential_breaking

    def build_prompt(self, commits: str) -> str:
        """
        Build the prompt for AI API

        Args:
            commits: Commit history

        Returns:
            Formatted prompt
        """
        prompt = (
            "Analyze the following Git commit history and generate a clean, structured changelog.\n"
            "Format: hash | author | date | message\n\n"
            "IMPORTANT: Detect Breaking Changes by looking for:\n"
            "- Commits with 'BREAKING CHANGE:', 'breaking change:', or '!' in commit messages\n"
            "- API changes that break backward compatibility\n"
            "- Removed features or deprecated functionality\n"
            "- Changes to data structures, configuration formats, or interfaces\n"
            "- Database schema changes requiring migration\n"
            "- Library replacements (e.g., Paper to MMKV, SharedPreferences to MMKV)\n"
            "- Storage migration or data format changes\n"
            "- Package renames or major refactoring\n"
            "- Commits with 'migrate', 'migration', 'replace X with Y' that affect user data\n\n"
            "Output ONLY the categorized changelog in this exact format:\n\n"
            "## Summary\n"
            "A concise summary of the changes (approx. 140 characters).\n\n"
            "## ⚠️ Breaking Changes\n"
            "- [hash] message (detailed explanation of what breaks and migration steps)\n\n"
            "## Features\n"
            "- [hash] message\n\n"
            "## Bug Fixes\n"
            "- [hash] message\n\n"
            "## Performance\n"
            "- [hash] message\n\n"
            "## Documentation\n"
            "- [hash] message\n\n"
            "## Other\n"
            "- [hash] message\n\n"
            "Rules:\n"
            "1. Do NOT include any introductory text, notes, or explanations\n"
            "2. Start directly with the Summary\n"
            "3. Breaking Changes section should be placed FIRST after Summary if detected\n"
            "4. Only include changelog categories that have commits (Summary is mandatory)\n"
            "5. Keep commit messages concise\n"
            "6. Use markdown format only\n"
            "7. If no breaking changes found, omit the Breaking Changes section entirely"
        )
        return f"{prompt}\n\nCommit History:\n{commits}"

    def summarize_changelog(self, commits: str) -> str:
        """
        Call Gemini API to summarize commits

        Args:
            commits: Commit history

        Returns:
            Generated changelog

        Raises:
            RuntimeError: If API request fails
        """
        try:
            response = self.client.models.generate_content(
                model=self.model,
                contents=self.build_prompt(commits),
                config=types.GenerateContentConfig(
                    temperature=0,
                    top_p=1,
                    max_output_tokens=100000,
                    candidate_count=1,
                )
            )

            return response.text

        except Exception as e:
            raise RuntimeError(f"Gemini API request failed: {e}")

    def generate(self, repo_path: str = ".") -> str:
        """
        Main method to generate changelog

        Args:
            repo_path: Path to git repository

        Returns:
            Generated changelog
        """
        print("Fetching Git commit history...")
        latest_tag = self.get_latest_tag(repo_path)
        print(f"Latest tag found: {latest_tag}")

        commits = self.get_git_commit_range(repo_path, latest_tag, "HEAD")

        if not commits:
            raise ValueError("No commit history found")

        print(f"\nFetched commits:\n{commits}")

        # Detect breaking changes
        has_breaking = self.detect_breaking_changes(commits)
        if has_breaking:
            print("\n⚠️  WARNING: Potential BREAKING CHANGES detected in commits!")
            print("    Please review the changelog carefully.\n")
        else:
            print("\n✓ No breaking changes detected in commit messages.\n")

        print("Calling Gemini API for summarization...")

        summary = self.summarize_changelog(commits)

        # Verify if AI detected breaking changes
        if "Breaking Changes" in summary or "⚠️" in summary:
            print("\n⚠️  BREAKING CHANGES DETECTED IN CHANGELOG!")
            print("    This release contains breaking changes that may affect users.\n")

        print(f"\n=== Changelog Summary ===\n{summary}")

        return summary

    def write_changelog(self, content: str, output_file: str = "CHANGELOG.md"):
        """
        Write changelog to file

        Args:
            content: Changelog content
            output_file: Output file path
        """
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(content)
        print(f"\nChangelog written to {output_file}")


def main():
    """Main entry point"""
    try:
        generator = ChangelogGenerator()
        changelog = generator.generate(".")
        generator.write_changelog(changelog)

    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()
